SA-21 Added newly agreed upon style, guide and naming conventions (#9)
Ultimaker/Meta: PythonCodeConventions.md

Python Code Conventions
=======

General
---
Ultimaker adheres to the PEP8 coding guidelines as described here: https://www.python.org/dev/peps/pep-0008/,
with the following adjustments, exceptions and specific choices:

* Always use (4) spaces instead of tabs.
* Strings use double quotes to match C++ style.

> * Maximum line length is set to 120 characters. See the 'line breaking' sectoin below for specific suggested PEP8 style choices.

Author: Doh

> * Maximum line length is set to 120 characters. See the 'line breaking' sectoin below for specific suggested PEP8 style choices.

Author: I understand and can relate, but readability is _very_ hard to determine and check for and most of the time based on personal preference. Sure there are dead giveaways, but the pain is always with discourse on the nitty-gritty. Having strict limits, like every other agreement we make, will remove the need for possible endless/useless discussions.

> * Maximum line length is set to 120 characters. See the 'line breaking' sectoin below for specific suggested PEP8 style choices.

Reviewer: Not putting any particular hard limit there also removes the need for possible endless discussions, as we had previously. Putting a hard limit here now initiates a possible endless discussion (currently 3 comments and counting).

> * Maximum line length is set to 120 characters. See the 'line breaking' sectoin below for specific suggested PEP8 style choices.

Author: A possibility too, surely. Well then, if you discuss this with the entire architecture team and provide me with the outcome, I'll 'report' back to 'my' team.

> * Maximum line length is set to 120 characters. See the 'line breaking' sectoin below for specific suggested PEP8 style choices.

Reviewer: This discussion is public enough, so I'm hoping that those interested will chime in.

> * Maximum line length is set to 120 characters. See the 'line breaking' sectoin below for specific suggested PEP8 style choices.

Author: Removed max line length.

* String formatting, multiple elements:
``` python

> "{foo}_{bar}".format(foo="foo", bar="bar")

Author: Hm, I'll discuss this with (some members of) the team since this differs from PEP8 but is not explicitly mentioned in the list of exceptions. Taken from PEP8:

`
Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.

Yes:

def complex(real, imag=0.0):
    return magic(r=real, i=imag)
No:

def complex(real, imag = 0.0):
    return magic(r = real, i = imag)
`

Also the "code_conventions.md" file is IMHO not a very good guide:
* Tries to copy PEP8, but is far from complete. Where I think we should not copy PEP8 at all, just the exceptions and specific choices we make therein.
* It is filled with TODOs and gaps.

> "{foo}_{bar}".format(foo="foo", bar="bar")

Author: Spaces were not discussed, team was under the impression we were using PEP8 on this. Will have to await next meeting.

> "{foo}_{bar}".format(foo="foo", bar="bar")

Reviewer 2: It has gasp todos that are also not handled in this document, so that shouldn't be the point. 

Going back onto discussion; My dyslexia really likes proper room around parts of sentences. This means that it chokes up on things like some_variable_that_we_need="someReallyLongButExplicitThing".

This is why I argued back in the day that we should break with PEP on that regard (It not being part of the python code conventions is intentional!)

> "{foo}_{bar}".format(foo="foo", bar="bar")

Author: The intention of this document was not to fill all gaps or missing guides. It addresses a few specific cases as discussed in our team's architecture meeting.

Doesn't matter what we do concerning the spaces, I was just relating to how it got to be like it is in this PR. Doesn't matter what we use, as long as we stick to it :) I can see how not using spaces, make reading harder for someone with dyslexia. Point made, I will 'report' that back to the team and adjust examples accordingly.

> "{foo}_{bar}".format(foo="foo", bar="bar")

Author: @nallath @Ghostkeeper added spaces, removed code_convention overlapping items. Will update/expand code_conventions.md where applicable.

```
* String formatting, single element:
``` python
"{}_bar".format("foo")
```
* Naming convention(s):

>  * PEP8 specifies functions to be in lower_case_with_underscores. At Ultimaker we decided to match our C++ coding convention and use mixedCase (CapitalizedWords by initial lowercase character)

Reviewer 2: This is duplicated from the code_conventions.md

Author: As it was before. But will (re)move this.

Author: Removed

>  * Parameters, variables and members follow the lower_case_with_underscores convention. PEP8 has no recommendation on this.

Reviewer 2: Why does it matter that pep does/does not have something to say about this?

Reviewer 2: Also, i think we should use the same style as the code convention document (eg; MUST, SHOULD, SHOULD NOT, etc.) That way it's clear what we mean by this.

Author: 1) to make a distinction between a deviation from PEP8 or a specific choice by Ultimaker without any explicit PEP8 'support'. This text was taken as-is from the previous version.
2) Good idea, will rewrite accordingly.

Author: Rewritten.

* Avoid circular dependencies as much as possible. Remove when encountered according to boy-scout rules.

> * Check the code_conventions file for more (generic) rules.

Reviewer 2: This should be the first thing the document states; All rules defined in the code_conventions file are used, except when explicit exceptions are stated.

Author: Will do

Python Code Conventions
=======

General

> ---

Reviewer 3: This document is mixing two markdown styles for the headers. Please make whole document consistent to use the newer style wit h'#', '##', etc.

Author: Will do

> Ultimaker adheres to the PEP8 coding guidelines as described here: https://www.python.org/dev/peps/pep-0008/,

Reviewer 3: ```suggestion
Ultimaker adheres to the PEP8 coding guidelines as described here: https://www.python.org/dev/peps/pep-0008/
```

> ** Important: **

Reviewer 3: ```suggestion
**Important:**
```

> Also check the code_conventions.md file for the generic rules, regardless of language.

Reviewer 3: ```suggestion
Also check the generic rules stated in the code_conventions.md document. These overrule PEP8, meaning Ultimaker deviates from the rules regarding function naming, documentation, etc.
```

With the following adjustments, exceptions and specific choices:
* See the 'line breaking' section below for specific suggested PEP8 style choices where function calls and definitions exceed the soft line length limit of 120 characters.
* MUST apply the following string formatting:
  * Multiple elements:
``` python

> "{foo}_{bar}".format(foo = "foo", bar = "bar")

Reviewer 4: I have to say that I've always preferred to have keyword arguments written with no extra spaces. I like it because it visually groups each argument as a compact unit and the spaces between args neatly separate them.

> "{foo}_{bar}".format(foo = "foo", bar = "bar")

Author: @ckielstra @ChrisTerBeke @sedwards2009 Although I tend to agree with you on not adding the extra spaces, they were added to support increased readability for dyslexic people of which Ultimaker employs quite a few. This was mentioned in earlier comments, but probably drowned with a lot of other comments :)

> "{foo}_{bar}".format(foo = "foo", bar = "bar")

Reviewer 5: And the problem with the spaces is that pep8 is unclear on what should be done. I personally think without spaces is more readable for my dyslectic mind. (as spaces separate the arguments).

> "{foo}_{bar}".format(foo = "foo", bar = "bar")

Author: @daid: formatting and logging; yes, this was not intended for use with logging, for the reasons you provided. Will add that to the description.
Concerning spaces: argl, so one says spaces help and the other says no spaces help... I'll leave it up to you (@daid @nallath) to figure something out :)

> "{foo}_{bar}".format(foo = "foo", bar = "bar")

Reviewer 6: I don't really care. In cloud we use spaces because that's what happens in Cura (so I took along the code style). We run flake8 for checking the code formatting in an automated way, so we can simply change the rules if something is decided here.

> "{foo}_{bar}".format(foo = "foo", bar = "bar")

Author: @ChrisTerBeke Let's stick to PEP8 then.

```
 * Single element:
``` python
"{}_bar".format("foo")
```

> * SHOULD avoid circular dependencies as much as possible. Remove when encountered according to boy-scout rules.

Reviewer 3: Avoiding circular dependencies is a generic good coding practice. I don't understand why it is mentioned at all and when it would be important to mention it should be in the code_conventions.md document and not here as an exception from the generic rule.

Author: This was discussed and suggested within a Python context only. I sure can see this as a general rule, but I'm not familiar with other languages covered by the generic guidelines to be able to judge if this holds up for all affected languages. @ChrisTerBeke can you?

Reviewer 6: I think this counts for every language. At least it also does for JavaScript/TypeScript (the only other language next to Python we use in Cloud and Connect).

Author: @ChrisTerBeke I believe this collides with forward declarations in Cpp? I'd like to stick to Python rules and not open up another can of worms a.t.m.

* Parameters, variables and members MUST follow the lower_case_with_underscores convention.

Type checking
---

> As per Ultimaker RFC-1,all new and changed code MUST be typed by using mypy typing. See: http://mypy-lang.org/ for more information.

Reviewer 3: ```suggestion
As per Ultimaker RFC-1, all new and changed code MUST be typed by using mypy typing. See: http://mypy-lang.org/ for more information.
```

In this case, mypy will inform that the call using the floating point value is not valid:
__mypytest.py:10: error: Argument 1 to "fib" has incompatible type "float"; expected "int"__

The following construct is _only_ allowed in case of unavoidable circular dependencies, otherwise regular import statements MUST be used.
``` python

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from [package] import [class]
```
## Line breaking
Indentation of function definitions and function calls that exceed the soft line length (PEP8: 80, Ultimaker: 120 characters) is described in PEP8 https://www.python.org/dev/peps/pep-0008/#indentation. From the suggested set, the following SHOULD be applied:

When a function definition exceeds the maximum line length, wrap the lines as such:
``` python

> def long_function_name(

Reviewer 7: I don't mind keeping it like this, but would slightly prefer putting the first bracket `(` on a newline, because that is congruent to how `{` is treated when starting blocks (function, if, etc..).

Reviewer 7: Just realized that this wouldn't really work for Python because you'd have to add line breaks every time. So nevermind :-)

In this case, mypy will inform that the call using the floating point value is not valid:
__mypytest.py:10: error: Argument 1 to "fib" has incompatible type "float"; expected "int"__

The following construct is _only_ allowed in case of unavoidable circular dependencies, otherwise regular import statements MUST be used.
``` python

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from [package] import [class]
```
Line breaking
---

> Indentation of function definitions and function calls that exceed the soft line length (PEP8: 80, Ultimaker: 120 characters) is described in PEP8 https://www.python.org/dev/peps/pep-0008/#indentation. From the suggested set, the following SHOULD be applied:

Reviewer 3: Don't hide the 120 character specification inside a paragraph but make it standout on its own line.

Author: It has his own line in the code_conventions.md, stated here as a quick reference. Can either stay this way or be removed entirely.

When a function definition exceeds the maximum line length, wrap the lines as such:
``` python
def long_function_name(
    var_one: str,
    var_two: int,
    var_three: float,
    *,
    var_four: bool = False
) -> None:
    print(var_one)
```
for a function call:
``` python
foo = long_function_name(var_one, var_two,

>     var_three, var_four = True

Author: Won't fix, see earlier explanation.

>     var_three, var_four = True

Author: Removed spaces according to PEP8

)
```
When encountering a long function call with named parameters, a newline MUST be used for each named parameter:
``` python
long_object_name.very_long_function_name(

>     this_is_the_first_variable = var_one,

Author: Won't fix, see earlier explanation.

>     this_is_the_first_variable = var_one,

Author: Changed to PEP8

>     this_is_the_second_variable = var_two

Author: Won't fix, see earlier explanation.

>     this_is_the_second_variable = var_two

Author: Changed to PEP8

In this case, mypy will inform that the call using the floating point value is not valid:
__mypytest.py:10: error: Argument 1 to "fib" has incompatible type "float"; expected "int"__

The following construct is _only_ allowed in case of circular dependencies, otherwise use regular import statements.
``` python
MYPY = False
if MYPY:

>     from [namespace] import [class]

Reviewer: I greatly prefer this solution, because it is less magic:

    from typing import TYPE_CHECKING
    if TYPE_CHECKING:
        from [package] import [class]

Author: It is suggested to drop the mechanism entirely.

Reviewer: Not entirely. Currently it is suggested to drop the mechanism except in cases where you'll have a circular dependency. So if you drop it entirely then you'll either have circular dependencies and the code won't run, or you'll have to import it locally in a function which also violates the PEP8 rule to import everything at the top of the file.

Author: Yes, not entirely, bad choice of words. But in case of _absolutely unavoidable_ circular dependencies. Currently the Controller class is passed around everywhere, which isn't necessary in quite a lot of occurrences.  Using the above construct hides those dependencies and stand in the way of approving on this.

Reviewer 2: Why would we use those only in case of circular dependencies? The code doesn't actually *need* to import it. It just needs it for the hinting. I'd actually argue that if the code doesn't need it, you should use the "softer" typing.

Reviewer 4: Why does the distinction matter? Type hints are code too.

Author: I agree with @sedwards2009 to some degree.  At least in that imports necessary for typing are grouped with the 'regular' imports by default.

Reviewer: In either case, regardless of when it is used, I would still prefer to use the `TYPE_CHECKING` variable rather than the magic `MYPY` variable.

Reviewer 2: If it doesn't matter which one, why define that it should only be used for circular imports? Eg; Why should we *only* use them to prevent circular imports (and thus not for the case that I'm using them for?).
I have a slight preference to just leave the circular import statement out and only state that we should use the TYPE_CHECKING in case you only want to import for use with type checking (and don't enforce a reason why someone would want to do this)

Author: @nallath @Ghostkeeper I do agree that using the TYPE_CHECKING variant is the better option. I do not agree with allowing this for every import related to a type added just for type checking.

Reviewer 3: The Application team discussed this before and we decided that in our team we don't want to use the MYPY / TYPE_CHECKING imports, unless you have to.

Reasoning: We had code where the conditional import made sense, but later code changed and did use the library instead of only for type checking. Normally PyCharm warns you for this error, but that doesn't work when this trick is used. These bugs are only detected run time.

Advantage of using the TYPE_CHECKING construct is minimal and to my opinion doesn't outweigh the hard bugs it can hide.

Reviewer 5: Saying that you never will have circular dependencies is IMHO crazy. While should be avoided, there are cases where it makes sense, especially for type checking.

The TYPE_CHECKING flag was added later in the typing module, but it is already available in the older version we use in the printer firmware. Also, it's introduced to improve typechecking compatibility with different tools, instead of having a tool specific flag. So I would prefer to use that flag instead of the custom MYPY which locks to a specific tool.

Reviewer 6: It seems the concensus is to prefer TYPE_CHECKING over MYPY? If so, let's settle and focus on other issues.

Author: @ChrisTerBeke Seems so, resolving conversation

```
Line breaking
---
Indentation of function definitions and function calls that exceed the maximum line length (PEP8: 80, Ultimaker: 120 characters) is described in PEP8 https://www.python.org/dev/peps/pep-0008/#indentation. From the suggested set, the following must be applied:

When a function definition exceeds the maximum line length, wrap the lines as such:
``` python
def long_function_name(
  var_one: str,
  var_two: int,
  var_three: float,
  *,

>   var_four: bool=False

Author: See comment above.

>   var_four: bool=False

Author: Added spaces

  ) -> None:

>   print(var_one)

Reviewer: Code style: Indent with 4 spaces.

Author: Fixed

```
for a function call:
``` python
foo = long_function_name(var_one, var_two,

>   var_three, var_four=True)

Author: See comment above.

>   var_three, var_four=True)

Author: Added spaces

```
When using long named parameters, use a newline for each named parameter:
``` python
long_object_name.very_long_function_name(
    this_is_the_first_variable=var_one,

>     this_is_the_second_variable=var_two

Reviewer: This is indented inconsistently with the function call and function definition above. Should we indent with 2 spaces or 4?

Reviewer: And perhaps we have to write down what indentation to use here, too.

Author: Isn't it already agreed upon that we use 4 spaces for each indentation level?
PEP8 describes which indentation to use in the suggested choices.
I did make a mistake with the spacing/indentation in the previous example, which is fixed.

Reviewer: As long as it's consistent within the code style standard, that's okay with me. 4 indents it is.

>     this_is_the_second_variable=var_two

Author: See comment above.

>     this_is_the_second_variable=var_two

Reviewer 2: Define long.

>     this_is_the_second_variable=var_two

Author: As stated before: exceeding 120 characters.

>     this_is_the_second_variable=var_two

Author: Added spaces

In this case, mypy will inform that the call using the floating point value is not valid:
__mypytest.py:10: error: Argument 1 to "fib" has incompatible type "float"; expected "int"__

The following construct is _only_ allowed in case of circular dependencies, otherwise use regular import statements.
``` python
MYPY = False
if MYPY:
    from [namespace] import [class]
```
Line breaking
---
Indentation of function definitions and function calls that exceed the maximum line length (PEP8: 80, Ultimaker: 120 characters) is described in PEP8 https://www.python.org/dev/peps/pep-0008/#indentation. From the suggested set, the following must be applied:

When a function definition exceeds the maximum line length, wrap the lines as such:
``` python
def long_function_name(
    var_one: str,
    var_two: int,
    var_three: float,
    *,
    var_four: bool=False

> ) -> None:

Reviewer 2: This is ambiguous. Should we always put *every* new parameter on a newline once it's longer? Because that isn't the case with the function call example below.

Author: This example is taken from PEP8, with some minor addittions (the boolean trap prevention and typing). The difference is in splitting up parameters for a function _definition_ and function _call_.  See: https://www.python.org/dev/peps/pep-0008/#indentation
I recall that we'd want to prevent spreading parameters for a long function call, because that would hamper reading the code of a single function. But with a function definition this would not be the case and putting each param on a new line, would contribute to readability, also because of the typing we prescribe.

Reviewer 2: Well, then the official one for pep8 is ambiguous. 

As for adding bool traps, I'm all in favor of putting it in there, but make it explicit (eg; an actual rule and not only something to be gleaned from an example).

Author: The intention was not to prescribe boolean traps (or prevention thereof) but to provide an example with as much function definition elements as possible.

Author: And yes, PEP8 does not prescribe a single way of splitting up long lines.

Reviewer 4: To be honest I really don't care that much about how people split lines as long as a little bit of common sense is applied. It is possible to define rules for this kind of thing, but most people can never remember what these "lesser used" rules are. Also just from my experience with auto-format functions in IDEs, rules for line splitting/wrapping usually produce awful and unreadable results compared to someone just using their brain for each situation.

Author: Problem is that what is considered 'common sense' differs per reviewer. The specific newly added point where discussed because we had trouble agreeing and think we would benefit from a/any specific choice. This will help/support uniform reviews opposed to different comments from different reviewers.

Author: Also; not suggesting auto formatting either.

Reviewer 4: To put it bluntly, I don't believe we can a) make a rule that people can remember, and b) make a rule which gives good readable results most of the time.

Reviewer 2: I agree with @sedwards2009. The biggest problem is making a rule that is both sensible and will be actually remembered (and subsequently; used).

Author: Well, this _is_ taken from a bunch of PEP8 proposed solutions, so whether they are taken out of this document, they are still valid and should/could be used/applied. So that would make them sensible, because we adhere to PEP8. The problem is that PEP8 offers more than one solution, it is suggested to make a specific choice in the matter.

Author: Applied @Ghostkeeper 's suggestion
